### ISSUE TITLE
Middleware Management and Organization 

### ISSUE THREAD

**Issue by pedronasser** [mholt/caddy#36]
I know @mholt is already working on a new way to create and import caddy's middlewares. But I'm here to open a discussion about something related to that.

I would like to leave here some statements (my opinion) for each topic.

## Core middleware

- Each core middleware should have it's own repository on github.

- The core should, for default, import only **essential and very common middlewares**, example: gzip, browse, proxy, redirect, extension.

- Other non common middlewares could be installed using the middleware management command line tool.

## Middleware management

- The source file that imports and registers all middlewares `config/middleware.go`, should have only importing and registering instructions. This will allow us to create a command-line tool for the middleware management, that only generates this source file.

**Comment by nemothekid** [comment]
Not sure I agree on the core defaults. If the core only imports essential middlewares like gzip, browse, proxy, redirect, extension - why would anyone use caddy over nginx (which is better supported, has higher performance and more widely used)?

Consider someone who might use the `markdown` middleware. Assuming they are unwilling or unable to learn how to recompile caddy for this middleware the default caddy distribution would only allow support for reverse proxying to a markdown server, something nginx can already do. Infact if they are already using nginx + some backend (like most people probably do - running some node/rails/php-fastcgi markdown app behind apache or nginx), theres little incentive to replace nginx with Caddy given that nginx most likely does that job better. And if that user has to recompile caddy for that middleware that processes *must* be easier than just editing config files in nginx. 

Simply put, without the other batteries, there is no reason for someone to choose Caddy over nginx. With those other batteries, the value proposition is better - with Caddy there is no need to have 2 processes for simple sites - just use caddy can serve your site and also render/cache your markdown (*and* you also get ssl, http/2 and all those nice features, which beats just hosting your node app on port 80).

**Comment by mholt** [comment]
We're now looking at a way to make this possible using a hosted solution (so users don't have to worry about building it themselves).

**Comment by mholt** [comment]
We're still fine-tuning things but I think today we got something working to allow anyone to extend Caddy easily. Developers can also apply to have their own middleware offered right from the Caddy website. @pedronasser your feedback is welcome: https://github.com/caddyserver/caddydev - here or on Slack is fine.

Note that this capability is not quite ready for everyone to use. Anyway, since all we have left (I think) is details, I'm going to close this issue. Thanks for the feedback!
